AMM/contracts/AMM/src/main.sw

```diff
@@ -0,0 +1,66 @@
+contract;
+
+dep errors;
+dep events;
+dep utils;
+
+use errors::InitError;
+use events::{RegisterPoolEvent, SetExchangeBytecodeRootEvent};
+use libraries::{AMM, Exchange};
+use std::{constants::BASE_ASSET_ID, logging::log, storage::StorageMap};
+use utils::bytecode_root;
+
+storage {
+    /// The valid exchange contract bytecode root
+    exchange_bytecode_root: Option<b256> = Option::None,
+    /// Map that stores pools, i.e., asset identifier pairs as keys and corresponding exchange contract identifiers as values
+    pools: StorageMap<(ContractId, ContractId), ContractId> = StorageMap {},
+}
+
+impl AMM for Contract {
+    #[storage(read, write)]
+    fn initialize(exchange_id: ContractId) {
+        require(storage.exchange_bytecode_root.is_none(), InitError::BytecodeRootAlreadySet);
+        let root = bytecode_root(exchange_id);
+        storage.exchange_bytecode_root = Option::Some(root);
+        log(SetExchangeBytecodeRootEvent { root });
+    }
+
+    #[storage(read, write)]
+    fn add_pool(asset_pair: (ContractId, ContractId), pool: ContractId) {
+        require(storage.exchange_bytecode_root.is_some(), InitError::BytecodeRootNotSet);
+        require(storage.exchange_bytecode_root.unwrap() == bytecode_root(pool), InitError::BytecodeRootDoesNotMatch);
+
+        let exchange_contract = abi(Exchange, pool.into());
+        let pool_info = exchange_contract.pool_info();
+        let pair_matches_exchange_pair = (pool_info.asset_a == asset_pair.0 && pool_info.asset_b == asset_pair.1) || (pool_info.asset_a == asset_pair.1 && pool_info.asset_b == asset_pair.0);
+
+        require(pair_matches_exchange_pair, InitError::PairDoesNotDefinePool);
+
+        let ordered_asset_pair = if asset_pair.0.into() < asset_pair.1.into() {
+            asset_pair
+        } else {
+            (asset_pair.1, asset_pair.0)
+        };
+        storage.pools.insert(ordered_asset_pair, pool);
```

Comment:
This looks like a storage modification after an external contract call (please see more detail in [this blog post](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html).
This is reported by my experimental feature developed in https://github.com/FuelLabs/sway/pull/3168.
Is this something we want to change or is it totally fine in this case?

Reply:
You are correct that this violates CEI, but note that

```require(storage.exchange_bytecode_root.unwrap() == bytecode_root(pool), InitError::BytecodeRootDoesNotMatch);```

guarantees that the external contract does precisely what is expected. Looking forward to the compiler catching these automatically though! :-)

Reply:
Awesome! Thanks :)